test/tls-provider.c

```diff
@@ -839,18 +3555,74 @@ int tls_provider_init(const OSSL_CORE_HANDLE *handle,
                       void **provctx)
 {
     OSSL_LIB_CTX *libctx = OSSL_LIB_CTX_new();
-
-    if (libctx == NULL)
+    OSSL_FUNC_core_obj_create_fn *c_obj_create= NULL;
+    OSSL_FUNC_core_obj_add_sigid_fn *c_obj_add_sigid= NULL;
+    OSSL_PROVIDER *defaultnullprov = NULL;
+    PROV_XOR_CTX *prov_ctx = NULL;
+
+    /* Fake signatures use HMAC and thus need default provider in NULL context */
+    if (!OSSL_PROVIDER_available(NULL, "default")) {
+        defaultnullprov = OSSL_PROVIDER_load(NULL, "default");
+        if (defaultnullprov == NULL)
+            return 0;
+    }
+    prov_ctx = xor_newprovctx(libctx, defaultnullprov);
+    if (libctx == NULL || prov_ctx == NULL)
         return 0;
 
-    *provctx = libctx;
+    *provctx = prov_ctx;
 
     /*
-     * Randomise the group_id we're going to use to ensure we don't interoperate
-     * with anything but ourselves.
+     * Randomise the group_id and code_points we're going to use to ensure we
+     * don't interoperate with anything but ourselves.
      */
-    xor_group.group_id = randomize_tls_group_id(libctx);
-    xor_kemgroup.group_id = randomize_tls_group_id(libctx);
+    xor_group.group_id = randomize_tls_alg_id(libctx);
+    xor_kemgroup.group_id = randomize_tls_alg_id(libctx);
+    xor_sigalg.code_point = randomize_tls_alg_id(libctx);
+    xor_sigalg_hash.code_point = randomize_tls_alg_id(libctx);
+
+    /* Register BIO for use */
+    if (!xor_prov_bio_from_dispatch(in))
+        return 0;
+
+
+    /* Retrieve registration functions */
+    for (; in->function_id != 0; in++) {
+        switch (in->function_id) {
+        case OSSL_FUNC_CORE_OBJ_CREATE:
+            c_obj_create = OSSL_FUNC_core_obj_create(in);
+            break;
+        case OSSL_FUNC_CORE_OBJ_ADD_SIGID:
+            c_obj_add_sigid = OSSL_FUNC_core_obj_add_sigid(in);
+            break;
+        /* Just ignore anything we don't understand */
+        default:
+            break;
+        }
+    }
+
+    /* Register algorithms manually as add_provider_sigalgs is
+     * only called during session establishment -- too late for
+     * key & cert generation...
+     */
```

Comment:
...

Reply:
fixed in https://github.com/openssl/openssl/pull/19312/commits/0330203fc25c8b2d08c052ef8070ab6f25d17209